### PR TITLE
Bitcoin keyring

### PR DESCRIPTION
**Pull request by supermassive**
Resolves https://github.com/brave/brave-browser/issues/28465
Sec review: https://github.com/brave/security/issues/1194

Bitcoin keyring initial support. The goal was to support these test vectors https://github.com/bitcoin/bips/blob/master/bip-0084.mediawiki#test-vectors

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/internal/hd_key.cc
Comment (supermassive) [inline components/brave_wallet/browser/internal/hd_key.cc:107]: moved into anonymous namespace

File: components/brave_wallet/browser/internal/hd_key_unittest.cc
Comment (supermassive) [inline components/brave_wallet/browser/internal/hd_key_unittest.cc:44]: This is now needed only for tests to match test vectors. Most likely to be moved to utils in the future

File: components/brave_wallet/browser/keyring_service.cc
Comment (supermassive) [inline components/brave_wallet/browser/keyring_service.cc:760]: Just made it into return early form

File: third_party/bitcoin-core/BUILD.gn
Comment (supermassive) [inline third_party/bitcoin-core/BUILD.gn:44]: Adding bech32 strencodings (bech32::Encode and ConvertBits) for `HDKey::GetSegwitAddress` needs

File: components/brave_wallet/browser/keyring_service.h
Comment (supermassive) [inline components/brave_wallet/browser/keyring_service.h:238]: Not exposing it through mojo, still not sure what interface we'll need

**Comment by kdenhartog** [comment]
Let's get a security review issue open for this one so I can use the issue to track the status of this, but can take a look at this one this afternoon

File: third_party/bitcoin-core/BUILD.gn
Comment (darkdh) [inline third_party/bitcoin-core/BUILD.gn:44]: Shouldn't there be a build error complaining about use of exception? Which is why we added [this](https://github.com/brave/brave-core/pull/8243/commits/261db3836286af02200cce70481b89e9f85a50a6) to deal with it without specifying `//build/config/compiler:exceptions`

Comment (supermassive) [reply third_party/bitcoin-core/BUILD.gn]: Well, that passed CI and my local compilation. But I think I completely misinterpreted what was going on here, so reverted changes to `serialize.h`

File: components/brave_wallet/browser/internal/hd_key.cc
Comment (supermassive) [reply components/brave_wallet/browser/internal/hd_key.cc:39]: Looks like a false positive

Comment (fmarier) [reply components/brave_wallet/browser/internal/hd_key.cc:39]: @thypon That rule also looks broken. There are no signed values here that I can see.

Comment (supermassive) [reply components/brave_wallet/browser/internal/hd_key.cc:39]: Technically 78 itself is an int and hence signed. 
But that `if the value of the signed primitive cannot be represented using an unsigned primitive`
is not the case so rule does not apply here.

**Review by kdenhartog — COMMENTED** [review]
Just an initial review - I've still got to look through a few other files but wanting to get you some quick feedback for now

File: components/brave_wallet/browser/brave_wallet_p3a.cc
Comment (supermassive) [reply components/brave_wallet/browser/brave_wallet_p3a.cc:248]: Browser will crash. 

Without `= nullptr` and switch miss browser will also most likely crash, but possibly could report weird metric(in case some valid string pointer appears on stack)

File: components/brave_wallet/browser/internal/hd_key.h
Comment (supermassive) [reply components/brave_wallet/browser/internal/hd_key.h:65]: This one actually is used only in tests now.

File: components/brave_wallet/browser/internal/hd_key_ed25519.cc
Comment (supermassive) [reply components/brave_wallet/browser/internal/hd_key_ed25519.cc:125]: I actually copypasted that from HDKeyEd25519::GetBase58EncodedPublicKey() assuming it was correct

File: components/brave_wallet/browser/keyring_service.cc
Comment (onyb) [inline components/brave_wallet/browser/keyring_service.cc:153]: I think we should either call this `kBitcoinNativeSegwitKeyringId` or introduce a separate parameter indicating the type of Bitcoin keyring. Here are some others we'll need to implement:

1. Pay-to-Public-Key-Hash (aka legacy) => `m/44'/0'` (widely used)
2. Pay-to-Script-Hash (aka segwit)  => `m/49'/0'` (not very widely used)
3. Pay-to-Taproot => `m/86'/0'` (new and not widely used, but gaining adoption)

The keyringId (or separate parameter, if you decide so) should then determine the kind of address encoding scheme to use. For example, `m/84'/0'` uses `GetSegwitAddress`, but we should not be able to invoke this function for a taproot-based Bitcoin keyring?

File: components/brave_wallet/browser/internal/hd_key_unittest.cc
Comment (onyb) [inline components/brave_wallet/browser/internal/hd_key_unittest.cc:37]: I believe the version byte will depend on the keyring, which is not the case with other methods in this file. Maybe just define `0x80` as a named constant for now, since the value will be different for bitcoin forks like Dogecoin.

File: browser/brave_wallet/keyring_service_unittest.cc
Comment (onyb) [inline browser/brave_wallet/keyring_service_unittest.cc:4475]: Is having an address for the account necessary? Perhaps `btc_acc->xpub` makes more sense here.

What also makes sense (for a future PR) is to have a `btc_acc->GetFreshReceivingAddress()` which returns an address for the first receiving index that is not used already (i.e., has no inputs/outputs on it). This is the address that we might want to expose to the user for deposits.

File: components/brave_wallet/browser/keyring_service.cc
Comment (supermassive) [reply components/brave_wallet/browser/keyring_service.cc:153]: I was planning to add support for different derivation paths(and don't forget testnets) once we understand we actually need them and I understand how to implement them :). 
Right now I thinking on focusing on minimal working prototype and fix problems once they appear.

File: browser/brave_wallet/keyring_service_unittest.cc
Comment (supermassive) [reply browser/brave_wallet/keyring_service_unittest.cc:4475]: Yep, account address does not make sense for bitcoin. But we now have 1-to-1 Account<->Address relation invariant which should be somehow handled in future. I just cover that behavior in tests.

`btc_acc->GetFreshReceivingAddress()` - sure, something like that was planned when RPC stuff is ready :) Thanks!

File: components/brave_wallet/browser/internal/hd_key_ed25519.cc
Comment (darkdh) [reply components/brave_wallet/browser/internal/hd_key_ed25519.cc:125]: @kdenhartog that is intentional for us to catch unhandled errors and it is not calling rust `unwrap` but with `expect`
https://github.com/brave/brave-core/commit/9ae530a91777a74ae22e631ca73a913ad708f6dc#diff-137afd46973b9774a2b139da7f989776c8aeafa5684cd1e6951014366b695c20R28-R29 which was reviewed in
https://github.com/brave/security/issues/718#issuecomment-1018064995
And it is very unlikely to fail because the key is generated from [`generate_ed25519_extended_secrect_key_from_seed`](https://github.com/brave/brave-core/blob/2531782df4a916b84913b552084ff987660aeeda/components/brave_wallet/browser/internal/hd_key_ed25519.cc#L30-L32) and [`generate_ed25519_extended_secrect_key_from_bytes`](https://github.com/brave/brave-core/blob/2531782df4a916b84913b552084ff987660aeeda/components/brave_wallet/browser/internal/hd_key_ed25519.cc#L43-L45), both check `is_ok()`. Also guarded [`CHECK`](https://github.com/brave/brave-core/blob/2531782df4a916b84913b552084ff987660aeeda/components/brave_wallet/browser/internal/hd_key_ed25519.cc#L23) in constructor which is impossible to hit now but just to prevent someone from making bad changes in the future.

File: browser/brave_wallet/keyring_service_unittest.cc
Comment (darkdh) [reply browser/brave_wallet/keyring_service_unittest.cc:4475]: We can make the address in the `AccountInfo` a union just like `TxDataUnion` so each different coin can encapsulate their own specific data.

File: components/brave_wallet/browser/brave_wallet_p3a.cc
Comment (kdenhartog) [reply components/brave_wallet/browser/brave_wallet_p3a.cc:248]: Ok seems like this change is useful, but also seems like a crash here wouldn't be ideal. Could we do a nullptr check before the and throw an error instead/not log?

File: components/brave_wallet/common/brave_wallet.mojom
Comment (supermassive) [reply components/brave_wallet/common/brave_wallet.mojom:929]: Sure, should be 0. Thanks!

**Review by darkdh — APPROVED** [review]
lgtm! Please squash commits before merging, thx.

File: components/brave_wallet/browser/brave_wallet_p3a.cc
Comment (supermassive) [reply components/brave_wallet/browser/brave_wallet_p3a.cc:250]: Added DCHECKs https://github.com/brave/brave-core/pull/17154/commits/d9e99f617c206f518f4f48a9183404ed1f422b63
which are enough for this particular case

File: components/brave_wallet/browser/internal/hd_key_ed25519.cc
Comment (kdenhartog) [reply components/brave_wallet/browser/internal/hd_key_ed25519.cc:125]: Awesome, thanks for all the links to easily follow this. Definitely not concerned with this anymore

**Comment by StephenHeaps** [comment]
Is it expected to get an error about creating bitcoin keyring when creating a wallet with the feature flag enabled (iOS)?
`[BraveCore] [keyring_service.cc:803] > Check failed: bitcoin_keyring.`
I might be misunderstanding scope of the PR.

File: components/brave_wallet/browser/internal/hd_key.cc
Comment (thypon) [reply components/brave_wallet/browser/internal/hd_key.cc:39]: fixed in https://github.com/brave/security-action/pull/106

**Comment by supermassive** [comment]
> Is it expected to get an error about creating bitcoin keyring when creating a wallet with the feature flag enabled (iOS)? `[BraveCore] [keyring_service.cc:803] > Check failed: bitcoin_keyring.` I might be misunderstanding scope of the PR.

Fixed in https://github.com/brave/brave-core/pull/17154/commits/94b3ae883fb05ebec92c390f8ccdc087525ce45c#
Thanks!

File: components/brave_wallet/browser/internal/hd_key.cc
Comment (kdenhartog) [inline components/brave_wallet/browser/internal/hd_key.cc:36]:
```suggestion
// a constant defined by BIP-32 to be the key for HMAC.
// This is safe because the entropy of the derived key comes from the seed phrase
constexpr char kMasterSecret[] = "Bitcoin seed";
```

Comment (kdenhartog) [reply components/brave_wallet/browser/internal/hd_key.cc:36]: @supermassive can we merge this suggestion in even though it wasn't relevant to your PR?

I came across the name of this constant and was confused about it's purpose so figured this might be helpful even though it's not directly relevant to the PR.

File: components/brave_wallet/browser/bitcoin_keyring.cc
Comment (kdenhartog) [inline components/brave_wallet/browser/bitcoin_keyring.cc:50]: nit: This function looks very similar to the one below. Would it make sense to refactor them together so that they call an internal function called DeriveKey that takes in a parameter for the child index that's passed to DeriveNormalChild?

File: components/brave_wallet/browser/bitcoin_keyring.cc
Comment (kdenhartog) [inline components/brave_wallet/browser/bitcoin_keyring.cc:13]: nit: seems like we could refactor this and GetChangeAddress together as well.

**Review by kdenhartog — APPROVED** [review]
Overall, I'm happy with this PR. I added a few more comments for the last bits I still needed to finish up reviewing, but neither are security blockers.

**Review by StephenHeaps — APPROVED** [review]

File: components/brave_wallet/browser/bitcoin_keyring.cc
Comment (supermassive) [reply components/brave_wallet/browser/bitcoin_keyring.cc:50]: It makes sense to refactor new code when whole feature gets in production and receives some updates. Right now I'm not sure where it all we go. Maybe similarity goes away and refactoring should be reverted hence double work.
I'm personally for some copypasting unless there is 110% confidence that similar code needs merge.
I'll add a TODO.

File: components/brave_wallet/browser/internal/hd_key.cc
Comment (supermassive) [reply components/brave_wallet/browser/internal/hd_key.cc:36]: Thanks! But will comment in next related PR.
Want to merge this one now.

**Pull request merged**
